### PR TITLE
Limit rubygems.reverse_dependencies to latest version of dependant

### PR DESCRIPTION
**Pull request by sonalkr132**
`reverse_dependencies` already joins three tables(dependencies, versions
and rubygems), using `includes(:latest_version, :gem_download)` makes
the query very big and slow.
This is probably an awful solution but something needs to be done :sweat:. 95 percentile of `ReverseDependenciesController#index` is always 10+ seconds.

### CODE REVIEW

**Comment by indirect** [comment]
This seems like a good temporary fix. I'd like to benchmark the results of doing separate queries, rather than joins, and also get a good EXPLAIN of the query that's being run, to see if we are missing indexes that would speed things up.

**Comment by sonalkr132** [comment]
I hadn't benchmarked because time reported in my dev log had a huge variance. The query has awful warmup time and depends on the number of reverse dependencies the gem had. 

I have limited the query to pick only latest versions of dependants. On reverse_dependencies page, we show the latest version of gem anyway.
<br/>
screenshot from 2017-03-03 14-20-12

Previously it was joining over all versions of dependants and then aggregating over `versions.rubygem_id`. Planner was underestimating number of rows returned on join nodes:
```
# planner expected 27 rows however it got 25,803 rows.
->  Nested Loop  (cost=20.06..3139.51 rows=27 width=4) (actual time=14.230..587.781 rows=25803 loops=1)
# planner expected 30 rows however it got 27,217 rows.
->  Nested Loop  (cost=19.64..3124.66 rows=30 width=4) (actual time=14.194..107.322 rows=27217 loops=1)
```
Implementation on this PR is 2.3 to 2.6 times faster than existing implementation. [gist](fb049eed77c7910176e3726450943277) with benchmark report and explain analyse of queries.

**Comment by indirect** [comment]
I'm happy with this as a solution. @dwradcliffe, any thoughts?

File: app/models/rubygem.rb [hunk 1]
```diff
@@ -275,7 +263,17 @@ def protected_days
   end
 
   def reverse_dependencies
-    self.class.reverse_dependencies(name)
+    self.class.joins("inner join versions as v on v.rubygem_id = rubygems.id
+      inner join dependencies as d on d.version_id = v.id").where("v.indexed = 't'
+      and v.position = 0 and d.rubygem_id = ?", id)
```
Comment (arthurnn) [inline app/models/rubygem.rb]: So now we only revert the dependencies for latest version?
Should we change the method name? Or add docs?

Comment (sonalkr132) [inline app/models/rubygem.rb]: > So now we only revert the dependencies for latest version?

Yes. we show a gem as reverse dependency only if its latest version depends on any of the versions of the dependency. [This assertion](https://github.com/rubygems/rubygems.org/pull/1574/files#diff-ee96e2d5cc5ac47f847e19de030867cdR231) intends to document the case where the latest version of `gem_three` no longer depends on `dependency`.

> Should we change the method name? Or add docs?

I will update [guide](http://guides.rubygems.org/rubygems-org-api/#get---apiv1gemsgem-namereverse_dependenciesjson) when this gets merged :+1:

**Comment by indirect** [comment]
I think this is a good solution to the performance problems caused by looking up reverse dependencies. Thanks for working on this!

**Pull request merged**